### PR TITLE
[New features] Add multimax LM-head modulation

### PR DESCRIPTION
**Pull request by ZhouYuxuanYX**
### PR Category

Operator Mechanism
### PR Types

New features
### Description

Add multimax, a learnable SeLU-style modulation applied to LM-head logits before language-modeling softmax/cross-entropy. The feature is opt-in through `TransformerConfig.multimax`, with `None` as the default disabled value and [`lm_head`] enabling the wired path. This PR adds LM-head parameters, routes them through `LanguageLoss` and fused linear CE, extends the Triton CE path, and adds single-card tests for config validation and LM-head multimax behavior.
#### Reference
Zhou, Yuxuan, Mario Fritz, and Margret Keuper. "MultiMax: Sparse and Multi-Modal Attention Learning." *International Conference on Machine Learning*. PMLR, 2024.

### CODE REVIEW

**Comment by ZhouYuxuanYX** [comment]
Re: Codecov patch coverage 5.02% — this is expected and not actionable in the current CI.

The bulk of the missed lines (≈155 / 208) are in two files:

- `src/paddlefleet/triton_ops/fused_linear_cross_entropy/cross_entropy.py` (Triton kernel)
- `src/paddlefleet/triton_ops/fused_linear_cross_entropy/fused_linear_cross_entropy.py` (PyLayer + forward/backward orchestration)

These are GPU-only paths: the kernel JIT-compiles via Triton and runs on CUDA. They are covered by `TestFusedMultimaxNumerical` in `tests/single_card_tests/model/test_lm_head_multimax.py`, which is gated on `paddle.is_compiled_with_cuda()` and verifies:

- forward loss parity vs Python `linear → SeLU → cross_entropy` ref (incl. `ignore_index` sample),
- grad parity on `_input`, `weight`, `multimax_ranges`, `multimax_ts`,
- regression test for the freeze-backbone / train-head case (`x.stop_gradient=True`, `w.stop_gradient=False`) that the bot review flagged.

The Codecov runner appears to be CPU-only, so these tests are skipped and the changed lines show as uncovered. Lifting patch coverage to 90% on these files would require either a GPU CI runner or contrived CPU mocks that don't actually exercise the kernel — neither of which adds real signal.

The remaining ~50 lines (`lm_head.py`, `language_loss.py`, `transformer_config.py`) are partially covered by `TestMultimaxLMHead` and `TestMultimaxConfig`; the uncovered portions there are the branches that route into the GPU fused path.

Suggest treating this Codecov status as advisory for this PR, or adding `cross_entropy.py` / `fused_linear_cross_entropy.py` to the Codecov ignore list (consistent with how other Triton ops in this repo would need to be handled). Happy to send a follow-up PR adding the ignore rule if maintainers prefer.

**Comment by ZhouYuxuanYX** [comment]
## Incremental coverage verification (local)

Ran `python ci/merge_and_check_coverage.py coverage.xml --fail-under 90` after filtering `coverage.xml` by the PR diff vs `upstream/develop` via `ci/filter_coverage.py`.

| File | Covered / Total | Rate |
|---|---|---|
| `models/common/language_loss/language_loss.py` | 12 / 12 | 100.00% |
| `transformer/transformer_config.py` | 16 / 16 | 100.00% |
| `triton_ops/fused_linear_cross_entropy/cross_entropy.py` | 1 / 1 | 100.00% |
| `triton_ops/fused_linear_cross_entropy/fused_linear_cross_entropy.py` | 116 / 119 | 97.48% |
| `models/gpt/lm_head.py` | 52 / 55 | 94.55% |
| **Total** | **197 / 203** | **97.04%** ✅ (≥ 90%) |

### Note on Triton kernels

`liger_cross_entropy_kernel` and `liger_cross_entropy_multimax_kernel` are `@triton.jit` functions whose bodies compile to PTX, so `coverage.py`'s Python AST instrumentation cannot observe them executing on GPU. The kernel `def` lines are marked `# pragma: no cover`; numerical correctness is verified end-to-end by:

- `tests/single_card_tests/custom_ops/test_fused_linear_ce_accuracy.py`
- `tests/single_card_tests/model/test_lm_head_multimax_selu.py`
- `tests/single_card_tests/ai_edited_test/triton_ops/test_ai_fused_linear_cross_entropy.py`

File: src/paddlefleet/models/gpt/lm_head.py [hunk 1]
```diff
@@ -12,6 +12,53 @@
...
+# =============================================================================
+# Multimax lm_head support
+# -----------------------------------------------------------------------------
+# When TransformerConfig.multimax in {"lm_head", "all"}, GPTLMHead adds two
+# learnable [4]-shape parameters (multimax_ranges, multimax_ts) and applies a
+# SeLU-style segmented modulation to logits before softmax/cross-entropy:
+#
+#     logits = SeLU(logits, multimax_ranges, multimax_ts)
+#
+# Both params are initialized to zero, which makes SeLU the identity at
```
Comment (From00) [inline src/paddlefleet/models/gpt/lm_head.py]: 这个每次续训都重新从0开始吗

File: src/paddlefleet/models/gpt/lm_head.py [hunk 2]
```diff
@@ -29,6 +76,44 @@
 from paddlefleet.transformer.identity_op import IdentityOp
 
 
+def SeLU(x, ranges, ts):
```
Comment (From00) [inline src/paddlefleet/models/gpt/lm_head.py]:
```suggestion
def LearnableSeLU(x, ranges, ts):
```
这个命名非常容易与框架的[SelU](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/SELU_cn.html#selu)算子混淆。

File: src/paddlefleet/models/gpt/lm_head.py [hunk 3]
```diff
@@ -109,6 +233,33 @@ def _forward(self, hidden_states: paddle.Tensor):
...
+            # materializes a full [B*S, V] logits tensor. Also fire the same
+            # one-shot rank-0 [MULTIMAX-LMHEAD-APPLIED] grep banner used by
+            # the unfused path so observability is consistent.
+            if getattr(self, "use_multimax_lmhead", False):
+                if not getattr(self, "_multimax_applied_logged", False):
+                    try:
+                        _rank = paddle.distributed.get_rank()
+                    except Exception:  # pragma: no cover - defensive: fleet always inits before forward
+                        _rank = 0
+                    if _rank == 0:
```
Comment (From00) [inline src/paddlefleet/models/gpt/lm_head.py]: 可以不需要这个判断，一般我们习惯性所有rank都输出。

File: src/paddlefleet/models/gpt/lm_head.py [hunk 4]
```diff
@@ -127,10 +278,41 @@ def recompute_handler(hidden_states, weight):
...
+                        f"logits.shape={list(logits.shape)} "
+                        f"ranges={self.multimax_ranges.detach().cast('float32').numpy().tolist()} "
+                        f"ts={self.multimax_ts.detach().cast('float32').numpy().tolist()}"
+                    )
+                self._multimax_applied_logged = True
+            # Wrap SeLU in `recompute` to drop saved intermediates: only the
+            # input logits (already kept by upstream autograd) are saved; SeLU
+            # is re-run during backward. Avoids OOM on full-vocab logits when
+            # `fused_linear_ce_loss_chunk=0` (required by multimax).
+            from paddle.distributed.fleet.utils import recompute
```
Comment (From00) [inline src/paddlefleet/models/gpt/lm_head.py]: 不太建议在函数内做import，可以移到文件开头。

File: src/paddlefleet/models/gpt/lm_head.py [hunk 5]
```diff
@@ -127,10 +278,41 @@ def recompute_handler(hidden_states, weight):
...
+
+            logits = recompute(
+                SeLU, logits, self.multimax_ranges, self.multimax_ts
+            )
+
+        # Loss-path MD5 probe: lm_head weight and logits (debug-only, opt-in via env vars)
```
Comment (From00) [inline src/paddlefleet/models/gpt/lm_head.py]: 这句注释的意思是？

File: src/paddlefleet/models/gpt/lm_head.py [hunk 6]
```diff
@@ -193,9 +375,20 @@ def sharded_state_dict(
...
+        TP ranks (no shard axis); they are listed explicitly so flex-checkpoint
+        saves/loads them.
+        """
         state_dict = self.state_dict(structured_name_prefix="")
-        shard_rules = None if self.world_size == 1 else {"weight": 0, "bias": 0}
+        if self.world_size == 1:
+            shard_rules = None
+        else:
+            shard_rules = {"weight": 0, "bias": 0}
+            if getattr(self, "use_multimax_lmhead", False):
```
Comment (From00) [inline src/paddlefleet/models/gpt/lm_head.py]: 感觉这段是不是不需要加也可以，非切分参数应该不需要改`sharded_state_dict`

File: src/paddlefleet/models/gpt/lm_head.py (hunk 1 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/models/gpt/lm_head.py]: 不是的，只有冷启动才是，其实冷启动也可以不设为零，但是还没有做这么细粒度的初始化对照试验。

File: src/paddlefleet/models/gpt/lm_head.py (hunk 2 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/models/gpt/lm_head.py]: 我待会儿更新pr改成SegLU(简称来自于全名Segmented Linear Unit)

File: src/paddlefleet/models/gpt/lm_head.py (hunk 4 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/models/gpt/lm_head.py]: 已采纳，见[272c4eb](https://github.com/PaddlePaddle/PaddleFleet/pull/1218/commits/272c4ebaafdb6e9622e28dc1a7dba317e478d300)

File: src/paddlefleet/models/gpt/lm_head.py (hunk 5 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/models/gpt/lm_head.py]: 已删除

File: src/paddlefleet/models/gpt/lm_head.py (hunk 6 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/models/gpt/lm_head.py]: Done

File: src/paddlefleet/models/gpt/lm_head.py (hunk 3 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/models/gpt/lm_head.py]: Done

File: src/paddlefleet/models/gpt/lm_head.py [hunk 7]
```diff
@@ -127,36 +280,28 @@ def recompute_handler(hidden_states, weight):
...
-                f"[LOSS_PATH_MD5] rank={rank} lm_head_weight shape={list(self.weight.shape)} md5={w_md5}",
-                flush=True,
-            )
-            print(
-                f"[LOSS_PATH_MD5] rank={rank} lm_head_input shape={list(hidden_states.shape)} md5={h_md5}",
-                flush=True,
-            )
-            print(
-                f"[LOSS_PATH_MD5] rank={rank} lm_head_logits shape={list(logits.shape)} md5={l_md5}",
-                flush=True,
```
Comment (LiYuRio) [inline src/paddlefleet/models/gpt/lm_head.py:159]: 这段为什么要删除

File: src/paddlefleet/transformer/transformer_config.py [hunk 8]
```diff
@@ -1287,3 +1308,59 @@ def __post_init__(self):
...
+            if isinstance(_multimax, _ListConfig):
+                _multimax = _OmegaConf.to_container(_multimax, resolve=True)
+                self.apply_multimax = _multimax
+        except ImportError:
+            pass
+        # Allow yaml/json to leave the field unset, set to ``null``, pass an
+        # empty string, or pass an empty list -- all map to the canonical
+        # disabled sentinel ``None``.
+        if _multimax in ("", []):
+            _multimax = None
```
Comment (LiYuRio) [inline src/paddlefleet/transformer/transformer_config.py:1337]: 为什么这里要加这么多奇怪的逻辑

File: src/paddlefleet/transformer/transformer_config.py [hunk 9]
```diff
@@ -292,6 +292,27 @@ class TransformerConfig(ModelParallelConfig):
     multimodal_embedding: bool = False
     """Whether to use multimodal embedding."""
 
+    apply_multimax: list[str] | None = None
```
Comment (LiYuRio) [inline src/paddlefleet/transformer/transformer_config.py]: 建议这个名字叫multimax_modules，参考recompute_modules

File: src/paddlefleet/transformer/transformer_config.py (hunk 8 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/transformer/transformer_config.py:1337]: 这个是机器人review以后给的bug修复建议，然后按照它的建议加的

Comment (ZhouYuxuanYX) [reply src/paddlefleet/transformer/transformer_config.py:1337]: 参考这里：
"src/paddlefleet/transformer/transformer_config.py
            _multimax = [_multimax]
            self.apply_multimax = _multimax
        if _multimax is not None:
            if not isinstance(_multimax, list) or not all(
@PaddlePaddle-bot
PaddlePaddle-bot
21 minutes ago
🔴 Bug apply_multimax: [lm_head] 通过 YAML 入口会被这里拒绝。

load_yaml() 返回的是 OmegaConf.create(result)，YAML 列表值不是内置 list，而通常是 ListConfig；这里强制 isinstance(_multimax, list) 会把 PR 描述和字段注释中推荐的 apply_multimax: [lm_head] 判成非法，只能 bare string 路径侥幸可用。这样 opt-in feature 在主入口 --configs 里不可用。

建议修复方式：

在校验前把 OmegaConf 容器转成普通 Python 容器，例如 from omegaconf import ListConfig 后对 ListConfig 执行 list(_multimax)；
或者使用 OmegaConf.to_container(_multimax, resolve=True) 统一归一化；
补一个从 load_yaml() 到 core_transformer_config_from_args() 的回归测试，覆盖 apply_multimax: [lm_head]。“

File: src/paddlefleet/models/gpt/lm_head.py (hunk 7 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/models/gpt/lm_head.py:159]: Hi @LiYuRio，感谢 review！

这段是我之前为了对齐 fused / unfused 两条路径的数值结果加的临时 MD5 probe，靠 LOG_LAYER_MD5 / LOG_LOSS_MD5 环境变量开关。删除主要有两个考虑：

From00 在前一轮 CR 里建议把调试代码清理掉，对齐验证已经完成，留在主路径里价值不大。
即使开关关闭，这段代码也会出现在每次 forward 的热路径上；一旦开启，.cast("float32").numpy().tobytes() 会触发 D2H 拷贝和全量 hash，对训练性能影响比较明显，不太适合作为常驻代码。
如果后续仍有数值对齐排查的需求，我倾向于把它放到独立的 debug hook / 工具里按需启用，而不是放回 lm_head.py 主路径。

File: src/paddlefleet/transformer/transformer_config.py (hunk 9 above)
Comment (ZhouYuxuanYX) [reply src/paddlefleet/transformer/transformer_config.py]: 有道理，行，apply_multimax听起来像bool类型，那我去更新一下，感谢反馈

File: src/paddlefleet/models/common/language_loss/language_loss.py [hunk 10]
```diff
@@ -247,12 +247,18 @@ def forward_impl(self, logits: Tensor | tuple, labels: Tensor) -> Tensor:
...
             )
 
-            hidden_states, weight, bias = logits
+            hidden_states, weight, bias = logits[:3]
+            # Multimax lm_head fused path: GPTLMHead emits a 5-tuple
+            # (hidden_states, weight, bias, multimax_ranges, multimax_ts)
+            # so SegLU is applied inside the chunked CE kernel without
+            # materializing full [B, S, V] logits.
+            multimax_ranges = logits[3] if len(logits) > 3 else None
+            multimax_ts = logits[4] if len(logits) > 4 else None
```
Comment (LiYuRio) [inline src/paddlefleet/models/common/language_loss/language_loss.py:256]: 这里建议不要硬取，这种magic_number建议写成dict

Comment (ZhouYuxuanYX) [reply src/paddlefleet/models/common/language_loss/language_loss.py:256]: 感谢建议，这个问题我会在下一个update来完善，暂时先保留该版本加速合入流程

**Review by From00 — APPROVED** [review]

**Review by ForFishes — APPROVED** [review]
LGTM
